Rework

ieure opened the pull request:
WIP PR that hugely reworks gocsv.  Nearly everything but the unmarshalling code has been jettisoned in favor of more encapsulated mechanisms.

Commit:
Remove normalizer.

It’s not actually used anywhere.

Commit:
Slash 'n burn

- Remove most of `decode.go` & tests.
  - All interfaces (`Decoder`, `SimpleDecoder`, `CSVReader`), their implementations, and helpers are gone.
  - Nuke `readTo`, `readToWithErrorHandler`, `readEach`, `readEachWithoutHeaders`, and `readToWithoutHeaders`.  They’re overly complicated, duplicative, and unnecessary.
- Remove most `csv.go`.
  - Remove all the global configuration: `selfCSVWriter`, `selfCSVReader`, and their getters/setters.
  - Remove literal dozens of `MarshalBlizzardOfTypesAndOptions` / `UnmarshalBlizzardOfTypesAndOptions` helper functions.
  - Remove `ErrorHandler` type alias, it was only used by the deleted `decoder` code.
- Remove `safe_csv.go`.
  - All it did is wrap `csv.Writer` with a mutex, so you can have concurrent processes writing to the same output, which is a bad idea to start.
  - Just pass your computed values to a writer goroutine with a channel, that’s what they’re for.
  - Or add your own locks around the writer.
  - But this is a bad pattern, if you use it, you should feel bad & deal with the consequences.
  - It doesn’t belong in the library.
- Move `UnmarshalCSVWithFieldsSample`'s `UnmarshalCSVWithFields()` function into the same file as the struct is defined in, rather than the random-ass one it was inexplicably in.

- Remove `safe_csv.go`.  All it does is lock/unlock around

Commit:
More S&B

- Move `TagName` and `TagSeparator` to `reflect.go`, and make them consts.  I cannot conceive of a world where you’d want to configure these at all, much less do so with globals.

Commit:
Remove the last of the global configuration.

- Move `FailIfUnmatchedStructTags`, `FailIfDoubleHeaderNames`, and `ShouldAlignDuplicateHeadersWithStructFieldOrder` into `Unmarshaller`.
- Make `Unmarshaller` actually check `FailIfUnmatchedStructTags`, which it didn’t previously do.
- Rename `csv.go` to `doc.go`, since it has no code left.  Update copyright.

Commit:
Simplify `Unmarshaller`.

- Instead of passing `unmarshalRow()` a map to conditionally mutate, have it unconditionally populate & return a fresh map.
- This simplifies `ReadUnmatched`, since it can return `unmarshalRow`'s return value directly.
- It also simplifies `Read()`, since it can thunk to `ReadUnmatched` and discard one argument.

Commit:
Remove `createNewOutInner`.

Commit:
Remove `errorInterface` and `isErrorType`.

Commit:
Revert "Remove `createNewOutInner`."

This reverts commit 91ffd270071b25a810d2e5833aeee9ccfb57181e.

Commit:
Config, pass 1.

Commit:
Config, pass 2.

Commit:
Config, pass 3.

Commit:
Config, pass 4

Commit:
Factor a bit of `unmarshalRow()` into `createNew()`.

Commit:
polish

Commit:
Remove support for default values.

Commit:
Config: include `structInfo`.

Commit:
WIP line number in errors

Commit:
Build Marshaller.

- Move some helpers from `encode.go` to `reflect.go`.
- Remove `encode*.go`.
- Build out `Marshaller` type & add a simple test.

Commit:
Fix module declaration.

Commit:
Add a `Reader` interface.

Commit:
Remove unused import.

Commit:
Format config.go.

Commit:
Add some convenience functions.

Commit:
Rename.

Commit:
Delete more unused code.

Commit:
Fix typos.

Commit:
Update README.

ieure commented on unmarshaller.go:
```diff
@@ -1,117 +1,114 @@
+	return value, err
+}
+
+func ReadAllCallback(um *Unmarshaller, cb func(interface{}) error) error {
```
I put this in as a helper method because I wasn't convinced that the `*Unmarshaller` needed this in it, but I could be convinced to move it inside.  Gocsv had so many specialized things shoved everywhere that I think I'd rather start simple and only add stuff unless it's very warranted.

ieure replied:
I also hate the name.

flaviut replied:
this is basically the following Java streams code:

```java
um.map(val -> ...)
    .takeWhile(err -> err == nil)
    .collect(Collectors.toList())
    .get(0)
```

Can we somehow do this in Go, and possibly even in a generic helper outside the Commando package?

ieure replied:
No.  The Java code works because Java has a type system with generics.  Since Go lacks that, it cannot express the output type as anything but `[]interface{}`, meaning any caller would need to traverse it to do something useful.

The `ReadAllCallback` function forces callers to do that as part of the parsing, which saves you from repeating the loop over the slice, and allocating two slices when you only need one.

bencdelaney replied:
> it cannot express the output type as anything but []interface{}, meaning any caller would need to traverse it to do something useful.

Agreed returning a `[]interface{}` is sub-optimal, but I believe we could instead use an appropriately-typed slice, which we pass around as an `interface{}` if we:

1. Take a `holder` as an `interface{}` parameter, assert that it's a `[]T`, and append records to `holder` instead of returning them, or
2. Create our own `[]T`, to which we append records, then return as an `interface{}`.

ieure replied:
I like this suggestion, I'll see if I can get something reasonably working that implements it.

ieure commented on unmarshaller.go:
```diff
@@ -1,117 +1,114 @@
+		rec, err := um.Read()
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		}
```
Having mixed feelings about this as well.  A definite goal is to let the caller decide what to do about errors, which this … doesn't.  The easy answer is to have the callback be `func(interface{}, error) error` and have it return `error` if it wants to stop processing.

I don't love this, since it means the straightforward two-line append-and-return-nil type callbacks get bloated with boilerplatey error checking/logging.

Maybe this should have two callbacks, ala:

```go
func ReadAllCallback(um *Unmarshaller, onSuccess func(interface{}) error, onError func(error) error) error
``` 

I also don't love this, but it does have the nice property of shoving all the error handling stuff into reusable functions, ala:

```go
func AbortOnAnyError(e error) error {
    return e
}

func LogAndConitnue(e error) error {
    log.Error(e, "Failed to unmarshal")
    return nil
}
```

This is perhaps also an argument for moving `ReadAllCallback` to take an `*Unmarshaller` pointer receiver — the error handler could be stuck into the config and default to stopping on the first error, which would give reasonable out of the box behavior without needing fiddly code at the callsite.  The downside would be that `Config` would have a member that's useless for Marshallers.

Open to suggestions.

jsandler18 replied:
I don't think it makes sense to make store the error callback in the unmarshaller/config if you are passing the success callback.  I would think both should be passed in each time. 

I am still not clear on what kind of errors OnError would actually be handling?

ieure replied:
> I don't think it makes sense to make store the error callback in the unmarshaller/config if you are passing the success callback. I would think both should be passed in each time.

Reasonable, but it sure is fiddly.  So calls with this approach would look like:

```go
	func LogAndContinue(err error) error {
		log.Error(ctx, "Failed to unmarshal record", err, nil)
		return nil
	}

	func StopOnError(err error) error {
		return err
	}

	err = commando.ReadAllCallback(um,
		func(rec interface{}) error {
			records = append(records, rec.(ConcreteType))
			return nil
		}, LogAndContinue,
	)

	err = commando.ReadAllCallback(um,
		func(rec interface{}) error {
			records = append(records, rec.(ConcreteType))
			return nil
		}, StopOnError,
	)
```

`StopOnError` could be part of Commando, the log function could be put in our `lib/` somewhere, since it depends on our internal logging code.

> I am still not clear on what kind of errors OnError would actually be handling?

Failures to unmarshal a row of CSV into a struct, for example, because we expect a float, but the value is `"$100" ` or `50-` or some other invalid string.  This type of error handling would let us parse all the valid rows within a file, while skipping invalid ones — whereas we currently fail the entire file if a single record cannot be unmarshalled.

jsandler18 replied:
I understand.  What you have here doesn't look so terrible to me.

ieure replied:
I don't love it, but it gets the job done.  Pushed commits.

Commit:
Actually implement `wrapLine()`.

Commit:
Handle errors.

bencdelaney commented on README.md:
````diff
@@ -1,17 +1,12 @@
 =====
 
-```go get -u github.com/gocarina/gocsv```
+```go get -u github.com/ieure/commando```
````
This is no longer accurate

```suggestion
```go get -u github.com/evenco/commando```
```

bencdelaney commented on README.md:
```diff
@@ -38,7 +33,7 @@ import (
 	"os"
 
-	"github.com/gocarina/gocsv"
+	"github.com/ieure/commando"
```
```suggestion
	"github.com/evenco/commando"
```

erikh replied:
also in imports, this will bite you subtly later if you miss it

bencdelaney commented on config.go:
```diff
@@ -0,0 +1,97 @@
+	// error when there is an unmatched struct tag.
+	FailIfUnmatchedStructTags bool
+
+	// FailIfDoubleHeaderNames indicates whether it is considered an
+	// error when a header name is repeated in the csv header.
+	FailIfDoubleHeaderNames bool
+
+	// ShouldAlignDuplicateHeadersWithStructFieldOrder indicates
+	// whether we should align duplicate CSV headers per their
+	// alignment in the struct definition.
+	ShouldAlignDuplicateHeadersWithStructFieldOrder bool
```
Do we have a real-world use case for these?

ieure replied:
We don't use them at all, they're held over from gocsv.

bencdelaney replied:
I'd - naively, maybe - lean towards axing them. Duplicate headers seems like a reasonable thing to not support.

Though I imagine plenty of real-world CSV's aren't so reasonable.

ieure replied:
I agree, but I think that's something I want to look at later.  Probably also nested struct serialization.

bencdelaney commented on go.mod:
```diff
@@ -1,3 +1,3 @@
-module github.com/gocarina/gocsv
+module github.com/ieure/commando
```
```suggestion
module github.com/evenco/commando
```

erikh replied:
just catching these as I see them

bencdelaney commented on marshaller.go:
```diff
@@ -0,0 +1,64 @@
+	return (&Config{Holder: holder}).NewMarshaller(writer)
+}
+
+// NewMarshaller creates a marshaller from a csv.Writer.
```
Might be worth noting `NewMarshaller` also writes the header.

erikh commented on README.md:
````diff
@@ -1,17 +1,12 @@
 =====
 
-```go get -u github.com/gocarina/gocsv```
+```go get -u github.com/ieure/commando```
````
post-conversation reminder: use `evenco/commando`

ieure replied:
I fixed it out from underneath you.

erikh commented on config.go:
```diff
@@ -0,0 +1,97 @@
+
+type Config struct {
+	// Holder is the type of struct to marshal from/unmarshal info.
+	Holder interface{}
```
consider using reflect.Type here

ieure replied:
It [ends up a `reflect.Type`](https://github.com/evenco/commando/pull/1/files#diff-0e426a43248661127a0c0ee115aef7a1093b635f8993b3f7ebb1dd9f05b8f249R88), this is just sugar so as a user of the lib you can pass the struct instead of needing to fiddle with `reflect`.

erikh replied:
I noticed, thought it'd be worth suggesting as an external change; nbd

erikh commented on marshaller.go:
```diff
@@ -0,0 +1,64 @@
+	return m.writer.Write(m.config.structInfo.headers())
+}
+
+func (m *Marshaller) Write(record interface{}) error {
```
just a general thought while I see all these `interface{}`s pop up. Would it make more sense to have typed interfaces that were required to implement a Write() or similar function instead of passing blanket types?

erikh replied:
maybe something that instead of actually /writing/ it produced a normalized structure that this function could process. 
hear me out and I am aware I am possibly misunderstanding the intent, as well as simplifying a bit. (typing this directly in the edit box here, so be kind)

```go

type Row map[string]string; // probably something more effective to preserve ordering like an idxhash or simply this:

type Row struct {
    Columns []string
    Elements []string
}

type Record interface {
    Yield() (Row, error)
}

func (m *Marshaller) Write(record Record) {
    row := record.Yield();
    // most of what's left of your function here
}
```

ieure replied:
This is pretty close to what `encoding/csv` does already, and I believe there's another library that turns things into `map[string]string`.  We have a ton of code that relies on gocsv using reflection to write structs out as CSV, so adopting something like this would mean writing a bunch of manual conversion functions.

In short, I think that's a reasonable design for a generic library started from scratch, but my aim is to take the stuff in gocsv that we rely on and make it less annoying to use, which I think that'd compromise.

erikh replied:
ack!

Commit:
Change ownership.

erikh commented on reflect.go:
```diff
@@ -132,12 +137,29 @@ func getConcreteReflectValueAndType(in interface{}) (reflect.Value, reflect.Type
+	case reflect.Ptr:
+		return nil
 	}
+	return fmt.Errorf("cannot use " + t.String() + ", only slice or array supported")
```
use %q here and it will quote it for free

erikh commented on marshaller.go:
```diff
@@ -0,0 +1,64 @@
+	if err := ensureStructOrPtr(inType); err != nil {
+		return err
+	}
+	inInnerWasPointer := inType.Kind() == reflect.Ptr
```
the comparison is probably trivial, so maybe just store the result here for comparisons later. there are other ways Elem() can explode IIRC

erikh replied:
what I mean is, you later pass this value to a function as reduced to bool, so it can be processed. what I'm suggesting is that there may be other reasons for this that may not be as easy to express this way, so you may just want to store the result of `Kind()` and process that in your function instead.

erikh replied:
and tbh, I think Kind is just a struct deref and likely all this work is done at TypeOf/ValueOf time; makes me wonder if the call overhead is really worth this optimization.

erikh commented on sample_structs_test.go:
```diff
@@ -100,11 +97,33 @@ type InnerStruct struct {
+	case "foo":
+		u.Foo = value
+	case "bar":
+		i, err := strconv.Atoi(value)
```
strconv.Parse* is better than Atoi for a lot of good reasons, I'll leave the docs to explain this and assume if you're using it knowing that you have a good reason. just wanted to note it.

erikh commented on types.go:
```diff
@@ -242,7 +242,7 @@ func setField(field reflect.Value, value string, omitEmpty bool) error {
 	default:
 		// Not a native type, check for unmarshal method
-		if err := unmarshall(field, value); err != nil {
+		if err := unmarshal(field, value); err != nil {
```
You can take the craft store out of a program, but you can't take the craft

erikh reviewed (commented):
I was asked by @ieure to review as a friend who can do golang things. Here are my comments and summary:

First, I really like what you've done here. Definitely a lot of cleanup.

Personally, if I needed to consume this library I would be worried about the large quantity of `interface{}` it requires, as type assertions and other muckery tend to make things hard to debug. I made some suggestions on how to curb some of this and I encourage you to at least consider them.

Two things I would like you to consider is more liberal use of `%w` (error wrapping) and `%q` (automatic quoting) in format strings. Most of my fights with CSV files in the past have been around quoting; making it easier for the user to determine that yes, their strings have issues due to quoting might be a worthwhile thing in this library.

Thanks again for the opportunity to help, and I hope my comments are worthwhile.

bencdelaney commented on unmarshaller.go:
```diff
@@ -1,117 +1,120 @@
+func ReadAllCallback(um *Unmarshaller, onSuccess func(interface{}) error, onError func(error) error) error {
+	for {
+		rec, err := um.Read()
+		if err == io.EOF {
```
Minor: Might want to use `errors.Is`, here.

bencdelaney commented on unmarshaller.go:
```diff
@@ -1,117 +1,120 @@
+func wrapLine(err error, line int) error {
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("On line %d: %w", line, err)
```
[Code review comments](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) would suggest lowercasing, here.

```suggestion
		return fmt.Errorf("on line %d: %w", line, err)
```

Alternatively, we could consider implementing our own error type here, which might let us do more interesting things in the `onError` callbacks.

bencdelaney commented on unmarshaller.go:
```diff
@@ -1,117 +1,120 @@
+	return value, err
+}
+
+// ReadAllCallback calls cb for every record Read() from um.
+func ReadAllCallback(um *Unmarshaller, onSuccess func(interface{}) error, onError func(error) error) error {
```
I'm not entirely sure what this gives us. I think we could pretty trivially implement this logic at callsites by successively calling `Read` directly until we hit `errors.Is(io.EOF)`, with no loss of flexibility.

This is pretty close to the API [`csv.Reader`](https://pkg.go.dev/encoding/csv#Reader) itself provides. Having a `ReadAll` helper might be convenient, but I'd lean towards returning a slice from `ReadAll` (potentially at the cost of some flexibility), and expect to do the looping ourselves when we need more control.

ieure replied:
> I'm not entirely sure what this gives us. I think we could pretty trivially implement this logic at callsites by successively calling `Read` directly until we hit `errors.Is(io.EOF)`, with no loss of flexibility.

I'm not sure I agree, but I'll see how things shake out after another pass or two at this.  I wrote this helper function after implementing the loop-and-handle pattern a couple times, and finding it unsatisfactory.

The callback specifically gives us:

1. No possibility of missing that you have to check for `io.EOF` and writing an accidental infinite loop.
2. `onError` allows reuse of error-handling strategies, instead of duplicating them inline everywhere we unmarshal.

Callers are still able to `Read()` in a loop where that makes sense.

> This is pretty close to the API [`csv.Reader`](https://pkg.go.dev/encoding/csv#Reader) itself provides. Having a `ReadAll` helper might be convenient, but I'd lean towards returning a slice from `ReadAll` (potentially at the cost of some flexibility), and expect to do the looping ourselves when we need more control.

It's intentionally close to (and relies on) `csv.Reader`'s API, except it pops out structs instead of `[][]string` which you have to manually process.

I explicitly chose not to implement a `ReadAll()` style function because these have been problematic for us in the past.  For one, we have files large enough that we can't hold the whole thing in memory, and have worked around that by breaking them up ahead of time — but incremental parsing is a better solution, because we don't *need* the whole thing in memory at once, ever.

Additionally, many of the places we parse unmarshal into `[]ConcreteType`, but return `[]InterfaceType`.  So even if we allocate an appropriate slice and return it as `[]interface`, callers would have to traverse it a second time to build the interface slice.

In short, while it seems like a simple and obvious thing to have, I think it has hidden drawbacks, so biased towards the eat-your-veggies approach of forcing incremental parsing, but making it a bit easier.

*But*, I also felt that having a solid base of 1:1 csv-record:struct would let us build that type of thing on top, if we need.

bencdelaney replied:
Yep - this all makes sense. Agreed that `ReadAll` is a bad approach for big files. A few more thoughts here, around a couple (potentially foolish) potential refinements:

> Additionally, many of the places we parse unmarshal into []ConcreteType, but return []InterfaceType.

I think we might be able to get around this if we pass in a `holder` slice, which could presumably be a slice of any type the concrete type in the config fulfills. There's a chance this would make the reflection bits way worse - I'm definitely not an expert on the details here.

> ...incremental parsing is a better solution, because we don't need the whole thing in memory at once, ever.

If we wanted to be even more clever, I could imagine passing in channel, instead of a slice, as the `holder`, then sending records down the channel as they're parsed.

Putting it all together, we could have something like:

```go
func (um *Unmarshaller) ReadAll(holder interface{}, onError ...func(error) error) error
```

where `holder` could be a `[]ConcreteType`, a `[]InterfaceType`, a `chan<- ConcreteType`, or a `chan<- InterfaceType`. Which would be a really concise API, at the cost of added internal complexity.

My gut feeling is that the channel bit is the bad kind of clever, unless it happens to be easier than I expect. The same is probably true about the `[]InterfaceType`.

bencdelaney replied:
Also, this is all to unmarshal a CSV, so we've likely already hit the point of diminishing returns in terms of API polish 🙃.

ieure replied:
Yeah, I agree that would be nice, but I also agree that it's complicated and unlikely to be used.  Gocsv has a bunch of this kind of stuff, which we have never needed.  But we do need a parser that can handle large & slightly broken files without blowing up.

bencdelaney reviewed (commented):
Added a few thoughts.

I think this looks solid, overall, though I think the callback approach might not add enough value to justify its somewhat unintuitive signature.

Commit:
Clean up `fmt.Errorf()`s.

Commit:
Remove whitespace.

Commit:
Address some review feedback.

Commit:
`ReadAll()` error handling.

Commit:
Wrap things up.

- Add some extra tests.
- Move `ReadAllCallback()` onto `Unmarshaller`.
- Add `Marshaller.WriteAll()` + test.
- Remove `ensureStructOrPtr()` and caller.  This is checked when the config is validated.
- Document.
